packages/andromeda_protocol/src/vault.rs

```diff
@@ -0,0 +1,210 @@
+use cw_storage_plus::Map;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+/// Mapping between (Address, Funds Denom) and the amount
+pub const BALANCES: Map<(String, String), Uint128> = Map::new("balances");
+pub const STRATEGY_CONTRACT_ADDRESSES: Map<String, String> =
+    Map::new("strategy_contract_addresses");
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub enum StrategyType {
+    Anchor,
+    NoStrategy,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct YieldStrategy {
+    pub strategy_type: StrategyType,
+    pub address: String,
+}
+
+impl StrategyType {
+    pub fn deposit(
+        &self,
+        storage: &dyn Storage,
+        funds: Coin,
+        recipient: &str,
+    ) -> Result<SubMsg, ContractError> {
+        let address = STRATEGY_CONTRACT_ADDRESSES.load(storage, self.to_string());
+        match address {
+            Err(_) => Err(ContractError::NotImplemented {
+                msg: Some(String::from("This strategy is not supported by this vault")),
+            }),
+            Ok(addr) => {
+                let msg = wasm_execute(
+                    addr,
+                    &ExecuteMsg::AndrReceive(AndromedaMsg::Receive(Some(to_binary(recipient)?))),
+                    vec![funds],
+                )?;
+                let sub_msg = SubMsg {
+                    id: 1,
+                    msg: CosmosMsg::Wasm(msg),
+                    gas_limit: None,
+                    reply_on: ReplyOn::Error,
+                };
+
+                Ok(sub_msg)
+            }
+        }
+    }
+}
+
+impl ToString for StrategyType {
+    fn to_string(&self) -> String {
+        match self {
+            StrategyType::Anchor => String::from("anchor"),
+            StrategyType::NoStrategy => String::from("none"),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct InstantiateMsg {
+    pub strategies: Vec<YieldStrategy>,
+    pub operators: Option<Vec<String>>,
+}
+
+impl InstantiateMsg {
+    pub fn validate(&self) -> Result<(), ContractError> {
+        for (idx, yield_strategy) in self.strategies.to_vec().iter().enumerate() {
+            if self.strategies.iter().enumerate().any(|(index, strategy)| {
+                index != idx
+                    && (strategy.strategy_type == yield_strategy.strategy_type
+                        || strategy.address == yield_strategy.address)
+            }) {
+                return Err(ContractError::StrategyNotUnique {
+                    strategy: yield_strategy.strategy_type.to_string(),
+                });
+            }
+
+            if yield_strategy.strategy_type == StrategyType::NoStrategy {
+                return Err(ContractError::InvalidStrategy {
+                    strategy: StrategyType::NoStrategy.to_string(),
+                });
+            }
```

Comment:
nit: Do this before the first check to fail early. Also if `NoStrategy` is invalid, can we just remove that variant?

Reply:
I think it could make sense to have a default strategy if `None` is passed for the `Option<YieldStrategy>` but I can see it going both ways. I think removing `NoStrategy` is fine for now as that feels a bit incompatible with how things are currently handled.